File: bk2ch10p541textLayoutGeometry/ch23p811selfSizingTextField/MyLayoutManager.m


#import "MyLayoutManager.h"

@interface MyLayoutManager ()
@end

@implementation MyLayoutManager

-(void)drawBackgroundForGlyphRange:(NSRange)glyphsToShow atPoint:(CGPoint)origin {
    [super drawBackgroundForGlyphRange:glyphsToShow atPoint:origin];
    if (!self.wordRange.length)
        return;
    NSRange range = [self glyphRangeForCharacterRange:self.wordRange actualCharacterRange:nil];
    range = NSIntersectionRange(glyphsToShow, range);
    if (!range.length)
        return;
    NSTextContainer* tc = [self textContainerForGlyphAtIndex:range.location effectiveRange:nil];
    CGRect r = [self boundingRectForGlyphRange:range inTextContainer:tc];
    r.origin.x += origin.x;
    r.origin.y += origin.y;
    CGContextRef c = UIGraphicsGetCurrentContext();
    CGContextSaveGState(c);
    CGContextSetStrokeColorWithColor(c, [UIColor blackColor].CGColor);
    CGContextSetLineWidth(c, 1.0);
    CGContextStrokeRect(c, r);
    CGContextRestoreGState(c);
}

@end

File: bk2ch10p541textLayoutGeometry/ch23p811selfSizingTextField/MyLayoutManager.swift

import UIKit

class MyLayoutManager : NSLayoutManager {
    
    var wordRange : NSRange = NSMakeRange(0,0)
    
    override func drawBackgroundForGlyphRange(glyphsToShow: NSRange, atPoint origin: CGPoint) {
        super.drawBackgroundForGlyphRange(glyphsToShow, atPoint:origin)
        if self.wordRange.length == 0 {
            return
        }
        var range = self.glyphRangeForCharacterRange(self.wordRange, actualCharacterRange:nil)
        range = NSIntersectionRange(glyphsToShow, range)
        if range.length == 0 {
            return
        }
        if let tc = self.textContainerForGlyphAtIndex(range.location, effectiveRange:nil) {
            var r = self.boundingRectForGlyphRange(range, inTextContainer:tc)
            r.origin.x += origin.x
            r.origin.y += origin.y
            let c = UIGraphicsGetCurrentContext()
            CGContextSaveGState(c)
            CGContextSetStrokeColorWithColor(c, UIColor.blackColor().CGColor)
            CGContextSetLineWidth(c, 1.0)
            CGContextStrokeRect(c, r)
            CGContextRestoreGState(c)
        }
    }
    
    
}
